### DISCUSSION TITLE
app-data objects and clog-body

### DISCUSSION THREAD

**Discussion by nathanvy** [rabbibotton/clog#165] [Q&A] [answered]
In [Demo 4](https://github.com/rabbibotton/clog/blob/main/demos/04-demo.lisp#L15) there's a class created to hold data that gets passed around between functions:

```cl
(defclass app-data ()
  ((head
    :accessor head)
   (side
    :accessor side)
   (main
    :accessor main)
   (sysop
    :initform nil
    :accessor sysop)))
```

However let's suppose I want to persist a login session so that I can close the browser and come back without having to log in again.  `clog-auth` appears to support this by [putting the auth key into html5 local storage](https://github.com/rabbibotton/clog/blob/main/source/clog-auth.lisp#L58).  

But if I want to follow a Demo 4-style architecture where I use an app-data object to pass around the contents of menus and headers and stuff, must I also pass around `body` so I can query the local storage?

For example I could do something like this:

```cl
(defun login-form (app)
  (create-label (main app) :content "Username:")
  (create-br (main app))
  (let ((uname (create-form-element (main app) :text)))
    (create-br (main app))
    (create-label (main app) :content "Password:")
    (create-br (main app))
    (let ((pwd (create-form-element (main app) :text)))
      (create-br (main app))
      (set-on-click (create-button (main app) :content "Submit")
		    (lambda (obj)
		      (declare (ignore obj))
		      (if (auth-user (escape-string (value uname)) (escape-string (value pwd)))
			  (progn
			    (setf (role app) :user)
			    (route-content app "Home"))
			  (clog-web:clog-web-alert (main app) "Get rekt" "Bad username and/or password." :time-out 3 :place-top t))))
      (set-on-click (create-button (main app) :content "Cancel")
		    (lambda (obj)
		      (declare (ignore obj))
		      (route-content app "Home"))))))
```

But I can't access `body` from there to make use of `clog-auth:store-authentication-token` unless I also add `body` as an argument.

Is there a way to access `body` without bringing it along?  Would it be the height of foolishness to store a reference to `body` as a member of class `app-data`?

**Comment by herbertjones** [comment, marked as answer]
It looks like the `clog-body` object is available by using `connection-body` on any `clog-obj`.

**Reply by rabbibotton** [reply]
That's the best way :)

**Reply by rabbibotton** [reply]
Here is a sample - (set-on-resize (window (connection-body obj)) #'on-size)

**Reply by nathanvy** [reply]
Thank you!

**Comment by aykaramba** [comment]
Thanks @herbertjones for the hint.  

For any newbs working with the pull down menuing templates that want to open up a popup and populate it with something that can draw to the main background window, you can just use something like   (create-div (connection-body obj) :content "TEST" :class "w3-red")  to access the body object.

Neat!
